mathesar/api/pagination.py

```diff
@@ -56,23 +62,76 @@ def paginate_queryset(
 
 class TableLimitOffsetGroupPagination(TableLimitOffsetPagination):
     def get_paginated_response(self, data):
-        return Response(OrderedDict([
-            ('count', self.count),
-            ('grouping', self.grouping),
-            ('results', data)
-        ]))
+        return Response(
+            OrderedDict(
+                [
+                    ('count', self.count),
+                    ('grouping', self.grouping),
+                    ('results', data)
+                ]
+            )
+        )
+
+    def iterate_filter_tree_and_apply_field_function(self, obj, field_function):
+        if type(obj) == list:
+            for filter_object in obj:
+                self.iterate_filter_tree_and_apply_field_function(filter_object, field_function)
+        elif type(obj) == dict:
+            if 'field' in obj.keys():
+                field_function(obj)
+            else:
+                for operator, filter_object in obj.items():
+                    self.iterate_filter_tree_and_apply_field_function(filter_object, field_function)
+
+    def extract_field_names(self, field_names):
+        def append_to_field_names(filter_obj):
+            if filter_obj['op'] == 'get_duplicates':
+                field_names.update(filter_obj['value'])
+            else:
+                field_names.add(filter_obj['field'])
+
+        return append_to_field_names
+
+    def convert_filter_field_id_to_name(self, column_map):
+        def x(filter_obj):
+            if filter_obj['op'] == 'get_duplicates':
+                filter_obj['value'] = [column_map[field_id].name for field_id in filter_obj['value']]
+            else:
+                field_id = filter_obj['field']
+                filter_obj['field'] = column_map[field_id].name
+
+        return x
 
     def paginate_queryset(
             self, queryset, request, table_id, filters=[], order_by=[], grouping={},
     ):
-        group_by = GroupBy(**grouping) if grouping else None
+        columns_ids = set()
```

Comment:
@silentninja I haven't looked at this code in a while, this logic is hard to follow, can you add comments to explain what you're doing?